pyteal/compiler/compiler.py

```diff
@@ -236,90 +250,267 @@ def sort_subroutine_blocks(
+
+        subroutineMapping: Dict[
+            Optional[SubroutineDefinition], List[TealComponent]
+        ] = sort_subroutine_blocks(subroutine_start_blocks, subroutine_end_blocks)
+
+        spillLocalSlotsDuringRecursion(
+            self.version, subroutineMapping, subroutineGraph, localSlotAssignments
+        )
 
-    subroutineLabels = resolveSubroutines(subroutineMapping)
-    teal = flattenSubroutines(subroutineMapping, subroutineLabels)
+        subroutineLabels = resolveSubroutines(subroutineMapping)
+        components: list[TealComponent] = flattenSubroutines(
+            subroutineMapping, subroutineLabels
+        )
 
-    verifyOpsForVersion(teal, options.version)
-    verifyOpsForMode(teal, options.mode)
+        verifyOpsForVersion(components, options.version)
+        verifyOpsForMode(components, options.mode)
 
-    if assembleConstants:
-        if version < 3:
-            raise TealInternalError(
-                f"The minimum program version required to enable assembleConstants is 3. The current version is {version}."
+        if self.assemble_constants:
+            if self.version < 3:
+                # TODO: convert this to a `verifyProgramVersion()` call
+                raise TealInternalError(
+                    f"The minimum program version required to enable assembleConstants is 3. The current version is {self.version}."
+                )
+            components = createConstantBlocks(components)
+
+        components = [TealPragma(self.version)] + components  # T2PT0
+        teal_chunks = [tl.assemble() for tl in components]
+        teal_code = "\n".join(teal_chunks)
+
+        full_cpb = _FullCompilationBundle(
+            ast=self.ast,
+            mode=self.mode,
+            version=self.version,
+            assemble_constants=self.assemble_constants,
+            optimize=self.optimize,
+            teal=teal_code,
+            teal_chunks=teal_chunks,
+            components=components,
+        )
+        if not with_sourcemap:
+            return full_cpb
+
+        source_mapper = _PyTealSourceMapper(
+            teal_chunks=teal_chunks,
+            components=components,
+            build=True,
+            teal_file=teal_filename,
+            include_pcs=pcs_in_sourcemap,
+            algod=algod_client,
+            annotate_teal=annotate_teal,
+            annotate_teal_headers=annotate_teal_headers,
+            annotate_teal_concise=annotate_teal_concise,
+        )
+        full_cpb.sourcemapper = source_mapper
+
+        # run a second time without, and assert that the same teal is produced
+        with sourcemapping_off_context():
+            assert NatalStackFrame.sourcemapping_is_off()
+            bundle_wo = self._compile_impl(
+                with_sourcemap=False,
+                pcs_in_sourcemap=False,
+                annotate_teal=False,
+                annotate_teal_headers=False,
+                annotate_teal_concise=True,
+            )
```

Comment:
nit: this part looks just like `compileTeal`

Reply:
a8e89f8135a885b05f78896b1764eb4c5057d531

Reply:
resolving